crates/sui-types/src/event.rs

```diff
@@ -1,22 +1,78 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use move_core_types::language_storage::StructTag;
+use move_core_types::language_storage::{ModuleId, StructTag};
+use name_variant::NamedVariant;
 use serde::{Deserialize, Serialize};
 use serde_with::{serde_as, Bytes};
 
-/// User-defined event emitted by executing Move code.
-/// Executing a transaction produces an ordered log of these
+use crate::{
+    base_types::{ObjectID, SequenceNumber, SuiAddress, TransactionDigest},
+    committee::EpochId,
+    messages_checkpoint::CheckpointSequenceNumber,
+};
+
+/// A universal Sui event type encapsulating different types of events
+#[derive(Debug, Clone, PartialEq)]
+pub struct EventEnvelope {
+    /// UTC timestamp in milliseconds since epoch (1/1/1970)
+    timestamp: u64,
+    /// Transaction digest of associated transaction, if any
+    tx_digest: Option<TransactionDigest>,
+    /// Specific event type
+    event: SuiEvent,
+}
+
+impl EventEnvelope {
+    pub fn new(timestamp: u64, tx_digest: Option<TransactionDigest>, event: SuiEvent) -> Self {
+        Self {
+            timestamp,
+            tx_digest,
+            event,
+        }
+    }
+
+    pub fn event_type(&self) -> &'static str {
+        self.event.variant_name()
+    }
+}
+
+/// Specific type of event
 #[serde_as]
-#[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
-pub struct Event {
-    pub type_: StructTag,
-    #[serde_as(as = "Bytes")]
-    pub contents: Vec<u8>,
+#[derive(Eq, Debug, Clone, PartialEq, NamedVariant, Deserialize, Serialize, Hash)]
+pub enum SuiEvent {
+    /// Move-specific event
+    MoveEvent {
+        type_: StructTag,
+        #[serde_as(as = "Bytes")]
+        contents: Vec<u8>,
+    },
+    /// Module published
+    Publish { package_id: ObjectID },
+    /// Transfer coin
+    TransferCoin {
+        object_id: ObjectID,
+        version: SequenceNumber,
+        destination_addr: SuiAddress,
+    },
```

Comment:
Publish and TransferCoin would correspond directly to transactions, right? Is there a reason to recapitulate the transaction information as an event, or should we just have a TransactionEvent that is published for each new transaction, which can then subsume these two (and many other cases such as SplitCoin, MergeCoin, etc)?

Reply:
The reason is that we want to have different event types with different fields, which we want to index.  For example, one might want to look up all transfers of a specific Object ID, or to a specific address.  I don't think we can have a TransactionEvent that can encompass all the different types, unless we had some kind of dynamic struct/map or something, at which point we might as well move towards using Move structs for everything (that would be one of the things we should discuss on here).